src/sentry/static/sentry/app/views/organizationEventsV2/transactionView/utils.tsx

```diff
@@ -0,0 +1,272 @@
+
+// get position of element relative to top/left of document
+const getOffsetOfElement = (element: Element) => {
+  // left and top are relative to viewport
+  const {left, top} = element.getBoundingClientRect();
+
+  // get values that the document is currently scrolled by
+  const scrollLeft = window.pageXOffset || document.documentElement.scrollLeft;
+  const scrollTop = window.pageYOffset || document.documentElement.scrollTop;
+
+  return {x: left + scrollLeft, y: top + scrollTop};
+};
+
+export const rectOfContent = (element: Element): Rect => {
+  const {x, y} = getOffsetOfElement(element);
+
+  // offsets for the border and any scrollbars (clientLeft and clientTop),
+  // and if the element was scrolled (scrollLeft and scrollTop)
+  //
+  // NOTE: clientLeft and clientTop does not account for any margins nor padding
+  const contentOffsetLeft = element.clientLeft - element.scrollLeft;
+  const contentOffsetTop = element.clientTop - element.scrollTop;
+
+  return {
+    x: x + contentOffsetLeft,
+    y: y + contentOffsetTop,
+    width: element.scrollWidth,
+    height: element.scrollHeight,
+  };
+};
+
+export const rectOfViewport = (): Rect => {
+  return {
+    x: window.pageXOffset,
+    y: window.pageYOffset,
+    width: window.document.documentElement.clientWidth,
+    height: window.document.documentElement.clientHeight,
+  };
+};
+
+export const rectRelativeTo = (rect: Rect, pos = {x: 0, y: 0}): Rect => {
+  return {
+    x: rect.x - pos.x,
+    y: rect.y - pos.y,
+    width: rect.width,
+    height: rect.height,
+  };
+};
+
+export const rectOfElement = (element: HTMLElement): Rect => {
+  const {x, y} = getOffsetOfElement(element);
+  return {
+    x,
+    y,
+    width: element.offsetWidth,
+    height: element.offsetHeight,
+  };
+};
+
+export const clamp = (value: number, min: number, max: number): number => {
+  if (value < min) {
+    return min;
+  }
+  if (value > max) {
+    return max;
+  }
+  return value;
+};
+
+export const isValidSpanID = (maybeSpanID: any) => {
+  return isString(maybeSpanID) && maybeSpanID.length > 0;
+};
+
+export const toPercent = (value: number) => {
+  return `${(value * 100).toFixed(3)}%`;
+};
+
+export type SpanBoundsType = {startTimestamp: number; endTimestamp: number};
+export type SpanGeneratedBoundsType =
+  | {type: 'TRACE_TIMESTAMPS_EQUAL'; isSpanVisibleInView: boolean}
+  | {type: 'INVALID_VIEW_WINDOW'; isSpanVisibleInView: boolean}
+  | {
+      type: 'TIMESTAMPS_EQUAL';
+      start: number;
+      width: number;
+      isSpanVisibleInView: boolean;
+    }
+  | {
+      type: 'TIMESTAMPS_REVERSED';
+      start: number;
+      end: number;
+      isSpanVisibleInView: boolean;
+    }
+  | {
+      type: 'TIMESTAMPS_STABLE';
+      start: number;
+      end: number;
+      isSpanVisibleInView: boolean;
+    };
+
+const normalizeTimestamps = (spanBounds: SpanBoundsType): SpanBoundsType => {
+  const {startTimestamp, endTimestamp} = spanBounds;
+
+  if (startTimestamp > endTimestamp) {
```

Comment:
Why is data like this valid?

Reply:
@lynnagara it's not valid. I'm swapping the timestamps so that the span bars can properly be rendered. 

I'm covering this case here just for completeness. 

I don't expect this to ever occur; but this should cover rare cases for when information is misreported from the SDKs. For these set of timestamps, a warning visual cue with a tooltip will be shown.

Reply:
Sounds good